Implemented user sessions

tusharad opened the pull request:
This PR introduces user session management for the Scotty, addressing the requirements outlined in issue #317. The implementation draws inspiration from the [Scotty Session library](https://hackage.haskell.org/package/scotty-session-0.0.5/docs/Web-Scotty-Session.html#t:ScottySM).

## Checklist:
- [x] Updated the changelog to reflect the new features.
- [x] Exposed the session management module via Web.Scotty.
- [x] Added comprehensive Haddock documentation for all public functions and types.
- [x] Included a practical example in the Example directory to demonstrate usage.
- [x] Developed test cases to ensure the robustness and reliability of the session management feature.

## Dependency Update: 
Added the `random` package to support the generation of session IDs. Please let me know if this addition is acceptable.

Do let me know if there are any further changes or additions you need. Thank you for reviewing this PR!

@ocramz

ocramz commented:
@tusharad Thank you! I think this is a very good PR. Just couple of minor remarks on documentation and a return type.

ocramz commented:
Hi @tusharad , hope you're doing well! I was wondering whether you need any help to land this PR? Please let us know :)

tusharad commented:
Hi @ocramz,

I believe there might have been a misunderstanding. Since you self-assigned this task, I assumed you would be implementing some of the changes yourself.

> "Just a couple of minor remarks on documentation and a return type."

Could you please elaborate on the specific return type and areas in the documentation you’d like to see revised? This will help me address your feedback more effectively.

Thanks!

ocramz commented:
Hi @tusharad , yes apologies about that!  

The return type I was mentioning is that of `getUserSession`, to distinguish the "no session" case from "expired". But I'm open to keeping this as it is with a strong enough argument :)

tusharad commented:
Here’s an improved and more polished version of your PR reply:

Hi @ocramz ,

Thank you for the feedback! In the latest update, I’ve replaced Nothing with Either and introduced a new type called SessionStatus, which can represent SessionExpired or SessionNotFound. Please let me know if there’s anything else you’d like to adjust or further refine. I appreciate your input!

ocramz commented on Web/Scotty.hs:
As far as I can see, this file (Web.Scotty) is missing the specialized versions of the `Session` functions, returning in the `ActionM` monad.  See e.g. https://hackage.haskell.org/package/scotty-0.22/docs/src/Web.Scotty.html#request

ocramz reviewed (changes requested):
A few outstanding comments, but mainly:
* missing ActionM specialized Session functions
* changelog entry with issue number
* Session module header with correct authors and copyright

Optional but nice to have : 
* add a mention of the new Session functionality in the README

tusharad commented:
Thanks for the review @ocramz!

- [x] Added ActionM specialized Session functions. I noticed the cookie functions don't have ActionM versions—should I open a separate issue for this?
- [x] Updated changelog entry with issue number
- [x] Updated session module header with correct authors and copyright.
- [ ] I missed mentioning sessions in README :( 
- [x] I also squeezed in optional expiration time as argument in createUsersSession.

Do let me know in anything else is required. Thanks again!

ocramz commented:
@tusharad Thanks for the revision! 

> the cookie functions don't have ActionM versions—should I open a separate issue for this?

Yes please, open a new ticket for this.

ocramz commented:
@tusharad thanks! There's still time to add a line to the README, in the Examples section :)

tusharad commented:
> There's still time to add a line to the README

@ocramz done :)

ocramz reviewed (approved):
Thank you @tusharad ! Great work. 👍
